src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+                                                          const ShellCommand& command)
+   {
+      boost::shared_ptr<ExecuteChunkOperation> pProcess =
+            boost::shared_ptr<ExecuteChunkOperation>(new ExecuteChunkOperation(
+                                                        docId,
+                                                        chunkId,
+                                                        command));
+      pProcess->registerProcess();
+      return pProcess;
+   }
+   
+private:
+   
+   ExecuteChunkOperation(const std::string& docId,
+                         const std::string& chunkId,
+                         const ShellCommand& command)
+      : isRunning_(false),
+        terminationRequested_(false),
+        docId_(docId),
+        chunkId_(chunkId),
+        command_(command)
+   {
+      using namespace core;
+      Error error = Success();
+      
+      // ensure staging directory
+      FilePath stagingPath = chunkOutputPath(
+               docId_,
+               chunkId_ + kStagingSuffix,
+               ContextExact);
+      
+      error = stagingPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = stagingPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // ensure regular directory
+      FilePath outputPath = chunkOutputPath(
+               docId_,
+               chunkId_,
+               ContextExact);
+      
+      error = outputPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = outputPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // clean old chunk output
+      error = cleanChunkOutput(docId_, chunkId_, true);
+      if (error)
+         LOG_ERROR(error);
+   }
+   
+public:
+   
+   ProcessCallbacks processCallbacks()
+   {
+      ProcessCallbacks callbacks;
+      
+      callbacks.onStarted = boost::bind(
+               &ExecuteChunkOperation::onStarted,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onContinue = boost::bind(
+               &ExecuteChunkOperation::onContinue,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onStdout = boost::bind(
+               &ExecuteChunkOperation::onStdout,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onStderr = boost::bind(
+               &ExecuteChunkOperation::onStderr,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onExit = boost::bind(
+               &ExecuteChunkOperation::onExit,
+               shared_from_this(),
+               _1);
+      
+      return callbacks;
+   }
+   
+private:
+   
+   enum OutputType { OUTPUT_STDOUT, OUTPUT_STDERR };
+   
+   void onStarted(ProcessOperations& operations)
+   {
+      isRunning_ = true;
```

Comment:
Agreed, I can't think of a reason for the isRunning flag.

On Thu, May 19, 2016 at 1:50 PM, Kevin Ushey <notifications@github.com>
wrote:

> In src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp
> <https://github.com/rstudio/rstudio/pull/706#discussion_r63924151>:
>
> > +
> > +      callbacks.onExit = boost::bind(
> > +               &ExecuteChunkOperation::onExit,
> > +               shared_from_this(),
> > +               _1);
> > +
> > +      return callbacks;
> > +   }
> > +
> > +private:
> > +
> > +   enum OutputType { OUTPUT_STDOUT, OUTPUT_STDERR };
> > +
> > +   void onStarted(ProcessOperations& operations)
> > +   {
> > +      isRunning_ = true;
>
> It's possible that we could just do without the isRunning_ flag (it looks
> like all we really need is the terminateRequested_ flag in this case)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/706/files/e45f8f2dd140bd229929d2f9ad0b6795f9366efe#r63924151>
>